lib/cylc/tests/test_config_upgrader.py

```diff
@@ -0,0 +1,114 @@
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+# Tests for the config upgrader - this is specifically for the function
+# testing that configs can be upgraded from Cylc 7 to 8.
+
+from parsec.OrderedDict import OrderedDictWithDefaults as ord_dict
+from cylc.cfgspec.suite import _upgrade_param_env_templates
+
+
+def test_upgrade_param_env_templates():
+    """Test that the deprecated [runtime][X][parameter environment templates]
+    contents are prepended to [runtime][X][environment], in the correct
+    order"""
+
+    def _cfg(dic):
```

Comment:
This looks like it should be a pytest fixture - I can draft something for you if you want.
The advantage of doing it as a fixture is that you could use

```python
@pytest.fixture
def _cfg(dic):
    ...
    return resutl

@pytest.mark.parameterize(
    'cfg, expected',
    [
        (

        ),
    ]
)
def test_upgrade_param_env_templates(_cfg, config, expected):
    assert  _cfg(config) == expected
```
The advantage of which is that each scenario shows up as a separate test, and can fail and be debugged separately.

Reply:
As the config returned by `_cfg()` is quite specific to this particular test, I'm not sure it's suitable to be made into a fixture?